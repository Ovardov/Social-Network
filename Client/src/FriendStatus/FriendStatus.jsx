import React, { Fragment, useContext } from 'react';
import userService from '../services/userService';
import { UserContext } from '../App/App';

function FriendStatus({ id, isFriends, props }) {
    const {username} =  useContext(UserContext);

    const handleAddFriend = () => {
        userService.addFriend(id)
            .then((res) => {
                if(res === 'Added Successfully') {
                    props.history.push({pathname: `/profile/${username}`, state: {showContentPage: 'Friends'}});
                }
            })
            .catch(err => console.log(err))
    }

    const handleRemoveFriend = () => {
        userService.removeFriend(id)
            .then((res) => {
                if(res === 'Removed Successfully') {
                    props.history.push({pathname: `/profile/${username}`, state: {showContentPage: 'Friends'}});
                }
            })
            .catch(err => console.log(err))
    }

    return (
        <Fragment>
            {isFriends === true
                ? <button className="button add-friend" onClick={handleRemoveFriend}>
                    <i className="fas fa-user-plus"></i>
                    Remove Friend
                  </button>

                : <button className="button add-friend" onClick={handleAddFriend}>
                    <i className="fas fa-user-plus"></i>
                    Add Friend
                  </button>
            }
        </Fragment>
    )
}

export default FriendStatus;