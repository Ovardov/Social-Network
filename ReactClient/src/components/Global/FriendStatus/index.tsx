// Libraries
import React, { FC as FC_, useState, useMemo } from 'react';
import { useSelector, useDispatch } from 'react-redux';
// Components
import Button from '../Buttons/Button';
// Services
import { addFriend, removeFriend } from '../../../services/userService';
// Utils
import { Colors } from '../../../utils/enums';
import { AppState as AppState_ } from '../../../redux';
import { UserState as UserState_, addFriendAction, removeFriendAction } from '../../../redux/actions/User';
// Models
import User_ from '../../../models/User';

interface Props {
  username: string
}

const FriendStatus: FC_<Props> = ({ username, }) => {
  const dispatch = useDispatch();
  const [isLoading, setIsLoading] = useState(false);

  const user = useSelector<AppState_, UserState_>(state => state.user);

  const isFriends = useMemo(() => {
    return user?.friends?.find((friend: User_) => friend.username === username);
  }, [user, username]);

  const isMyProfile = useMemo(() => {
    return user.username === username;
  }, [user, username]);

  const handleClick = async () => {
    setIsLoading(true);

    try {

      if (isFriends) {
        const { message, user: friend, } = await removeFriend(username) as { message: string, user: User_ };
        dispatch(removeFriendAction(friend));

        // ToDo - print message
        //console.log(message);
      }

      if (!isFriends) {
        const { message, user: friend, } = await addFriend(username) as { message: string, user: User_ };
        dispatch(addFriendAction(friend));

        // ToDo - print message
        //console.log(message);
      }
    } catch (err) {
      console.log(err);
    }

    setIsLoading(false);
  };

  if (isMyProfile) {
    return null;
  }

  return (
    <Button
      onClickHandler={handleClick}
      text={isFriends ? 'Remove Friend' : 'Add Friend'}
      color={Colors.PRIMARY}
      isLoading={isLoading}
    />
  );
};

export default FriendStatus;