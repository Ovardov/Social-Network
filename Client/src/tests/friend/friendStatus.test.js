import React from 'react'
import renderer from 'react-test-renderer';
import FriendStatus from '../../components/FriendStatus/FriendStatus';

describe('FriendStatus component', () => {
    test('should render remove button if is friends', () => {
        const component = renderer.create(
            <FriendStatus isFriends={true} />
        );

        expect(component.toJSON()).toMatchSnapshot();
    });

    test('should render add button if is not friends', () => {
        const component = renderer.create(
            <FriendStatus isFriends={true} />
        );

        expect(component.toJSON()).toMatchSnapshot();
    });
})
