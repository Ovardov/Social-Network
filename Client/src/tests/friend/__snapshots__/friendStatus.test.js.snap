// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FriendStatus component should render add button if is not friends 1`] = `
<button
  className="button add-friend"
  onClick={[Function]}
>
  <i
    className="fas fa-user-plus"
  />
  Add Friend
</button>
`;

exports[`FriendStatus component should render remove button if is friends 1`] = `
<button
  className="button add-friend"
  onClick={[Function]}
>
  <i
    className="fas fa-user-plus"
  />
  Remove Friend
</button>
`;
